### PR TITLE
8864 safe str unicode

### PR DESCRIPTION
**Pull request by markrwilliams**
http://twistedmatrix.com/trac/ticket/8864

Note this changes the API for `log.LogPublisher.msg` under Python 2.

### CODE REVIEW

File: src/twisted/python/reflect.py [hunk 1]
```diff
@@ -436,6 +436,13 @@ def safe_str(o):
             return o.decode('utf-8')
         except:
             pass
+    if not _PY3:
+        # On Python 2, attempt to encode a unicode representation
+        # first.
+        try:
+            return unicode(o).encode('ascii', 'backslashreplace')
+        except:
```
Comment (Julian) [inline src/twisted/python/reflect.py:444]: Exception? Although it seems the other except blocks here are already setting this precedent.

File: src/twisted/test/test_log.py [hunk 2]
```diff
@@ -390,20 +390,19 @@ def testMultipleString(self):
 
     def test_singleUnicode(self):
         """
-        L{log.LogPublisher.msg} does not accept non-ASCII Unicode on Python 2,
-        logging an error instead.
+        L{log.LogPublisher.msg} encodes Unicode as ``ascii`` with the
+        ``backslashreplace`` on Python 2.
```
Comment (Julian) [inline src/twisted/test/test_log.py]: I like the wording below better ("with `backslashreplace` error handling)

File: src/twisted/python/reflect.py (hunk 1 above)
Comment (markrwilliams) [reply src/twisted/python/reflect.py:444]: I guess it's important to catch `KeyboardInterrupt`?  I did leave the bare `except:` for consistency's sake.

File: src/twisted/test/test_log.py (hunk 2 above)
Comment (markrwilliams) [reply src/twisted/test/test_log.py]: Me too!  Should now be fixed

File: src/twisted/python/reflect.py (hunk 1 above)
Comment (Julian) [reply src/twisted/python/reflect.py:444]: Fair enough (sounds dubious, but hey, ain't your dubious)

**Pull request merged**